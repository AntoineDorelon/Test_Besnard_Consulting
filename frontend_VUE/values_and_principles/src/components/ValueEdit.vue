<template>
 <div v-if="value">
   <p>{{value.title.charAt(0).toUpperCase() + value.title.slice(1)}}</p>
 </div>
</template>

<script>
export default {
  name: "ValueEdit",
  props: ['value']
}
</script>

<style scoped>

</style>