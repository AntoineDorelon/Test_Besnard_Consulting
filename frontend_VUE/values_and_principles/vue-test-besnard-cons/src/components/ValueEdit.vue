<template>
  <div v-if="value">
    <p>
      <span style="text-decoration: underline">Value :</span> &nbsp;
      <label for="title">{{
        value.title.charAt(0).toUpperCase() + value.title.slice(1)
      }}</label>
      <br />
      <input id="title" placeholder="Title" v-model="localValue.title" />
      <button @click="saveValue()" type="button" class="btn btn-primary">
        Save the modifications
      </button>
    </p>
  </div>
</template>

<script>
export default {
  name: "ValueEdit",
  props: ["value"],
  data() {
    return {
      localValue: { ...this.value },
    };
  },
  watch: {
    value: function (newval, oldval) {
      if (newval !== oldval) {
        this.localValue = { ...this.value };
      }
    },
  },
  methods: {
    saveValue() {
      if (this.value.id) {
        console.log("Value to edit is :", this.value.title);
        fetch(`http://127.0.0.1:8000/api/values/${this.value.id}/`, {
          method: "PUT",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({ title: this.localValue.title }),
        })
          .then((response) => response.json())
          .then((response) => {
            this.$emit("ValueUpdated");
            console.log(response);
          })
          .catch((error) => console.log(error));
      } else {
        fetch(`http://127.0.0.1:8000/api/values/`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({ title: this.localValue.title }),
        })
          .then((response) => response.json())
          .then((response) => {
            this.$emit("ValueUpdated");
            console.log(response);
          })
          .then((this.value = { title: "" }))
          .catch((error) => console.log(error));
      }
    },
  },
};
</script>

<style scoped>
p {
  text-align: left;
}

button {
  margin-left: 2rem;
  margin-bottom: 1rem;
  margin-top: 1rem;
}
</style>
