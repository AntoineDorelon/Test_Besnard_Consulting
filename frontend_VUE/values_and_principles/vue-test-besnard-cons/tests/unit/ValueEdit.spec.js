import { mount } from "@vue/test-utils";
import ValueEdit from "../../src/components/ValueEdit";

describe("ValueEdit", () => {
  const value = { id: 1, title: "test-value" };
  const wrapper = mount(ValueEdit, {
    propsData: { value },
  });

  test("contains the value title in 'Camel case' when a value is passed", () => {
    expect(wrapper.html()).toContain(
      value.title.charAt(0).toUpperCase() + value.title.slice(1)
    );
  });

  test("contains the value in the placeholder", () => {
    // expect(wrapper.html()).toMatchSnapshot();
    const placeholderValue = wrapper.find("input").element.value;
    expect(placeholderValue).toBe("test-value");
  });

  test("calls saveValue() method when the button is clicked", async () => {
    const button = wrapper.find("button");
    const spy = jest.spyOn(wrapper.vm, "saveValue");
    await button.trigger("click");
    expect(spy).toHaveBeenCalled();
  });
});
