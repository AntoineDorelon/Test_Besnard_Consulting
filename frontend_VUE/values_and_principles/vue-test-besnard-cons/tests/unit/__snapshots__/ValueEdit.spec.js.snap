// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueEdit contains the value in the placeholder 1`] = `
<div>
  <p><span style="text-decoration: underline;">Value :</span> &nbsp;
    <label for="title">Test-value</label> <br> <input id="title" placeholder="Title"> <button type="button" class="btn btn-primary">
      Save the modifications
    </button>
  </p>
</div>
`;
